proposals/proposalTypes/Proposal.sol

```diff
@@ -1,110 +1,265 @@
+import {VmSafe} from "@forge-std/Vm.sol";
+import {console} from "@forge-std/console.sol";
 
-import {Core} from "@protocol/core/Core.sol";
-import {Roles} from "@protocol/core/Roles.sol";
+import {Script} from "@forge-std/Script.sol";
+import {IProposal} from "./IProposal.sol";
 import {Addresses} from "@proposals/Addresses.sol";
 
-abstract contract Proposal is IProposal, Test {
-    bool public DEBUG = true;
+abstract contract Proposal is Test, Script, IProposal {
+    using Strings for *;
+
+    struct Action {
+        address target;
+        uint256 value;
+        bytes arguments;
+        string description;
+    }
+
+    /// @notice starting snapshot of the contract state before the calls are made
+    uint256 private _startSnapshot;
+
+    /// @notice list of actions to be executed, regardless of proposal type
+    /// they all follow the same structure
+    Action[] public actions;
+
+    /// @notice debug flag to print proposal actions, calldata, new addresses and changed addresses
+    /// @dev default is true
+    bool internal DEBUG;
+
+    /// @notice Addresses contract
+    Addresses public addresses;
 
-    Core _core;
+    /// @notice the actions caller name in the Addresses JSON
+    string public caller;
 
-    function setDebug(bool value) external {
-        DEBUG = value;
+    constructor(string memory _caller) {
+        string memory environment = vm.envOr("ENVIRONMENT", string("localnet"));
+        string memory addressesPath = string(abi.encodePacked("proposals/Addresses/", environment, ".json"));
+        addresses = new Addresses(addressesPath);
+        caller = _caller;
+        DEBUG = vm.envOr("DEBUG", true);
     }
 
-    /// @notice run the deployment for testing only. ie intergration testing with foundry
-    /// @dev this is not run on-chain, and is only used with the foundry `forge test` command
-    function deployForTestingOnly(Addresses addresses, address deployer) public {
-        vm.startBroadcast(deployer);
+    /// @notice override this to set the proposal name
+    function name() public view virtual returns (string memory);
+
+    /// @notice override this to set the proposal description
+    function description() public view virtual returns (string memory);
 
-        _beforeDeploy(addresses, deployer);
-        _deploy(addresses, deployer);
-        _afterDeploy(addresses, deployer);
-        _aferDeployForTestingOnly(addresses, deployer);
+    /// @notice main function
+    /// @dev do not override
+    function run() external {
+        address deployer = addresses.getAddress("DEPLOYER");
 
+        vm.startBroadcast(deployer);
+        _beforeDeploy();
+        _deploy();
         vm.stopBroadcast();
 
-        _build(addresses, deployer);
-        _run(addresses, deployer);
+        address afterDeployCaller = addresses.getAddress("AFTER_DEPLOY_CALLER");
+        vm.startBroadcast(afterDeployCaller);
+        _afterDeploy();
+        vm.stopBroadcast();
```

Comment:
I have added `afterDeployCaller` in case we want to simulate actions that the deployer doesn't have permission to execute. For eg: On mainnet ztx deployer EOA deploys contracts but it doesn't have admin role access and cannot perform `_afterDeploy` actions.

For testnets `deployer` and `afterDeployCaller` could be the same but for mainnet `deployer` is ztx deployer contract and `afterDeployCaller` is admin multisig.